keps/draft-20180418-security-profile.md

```diff
@@ -0,0 +1,1516 @@
+- The `Deployment` objects to deploy controllers to enforce the rules
+  - Controllers watching namespaces and creates policy objects
+  - External admission control webhook
+  - Related configuration objects
+
+Please refer to [default-1.0-0](#the-default-profile) as an example.
+
+The Custom Resource `Security Profile` defines two sets of rules: bootstrapping rules and runtime rules
+
+#### Bootstrapping Rules
+
+The bootstrapping rules provide instructions for Kubernetes deployment tools
+to set up configurations correctly during cluster creating, upgrading/downgrading, including:
+
+- Command line flags for etcd, API server, controller manager, scheduler and kubelet
+- Locations and permissions of certificates, keys, configuration files
+- Security-relevant Operating System settings, for example
+  - Configurations for SELinux, AppArmor, seccomp
+  - Rules in iptables
+
+All of them are described as actionable rules which contains:
+
+- `name`: the unique name which identifies _a thing_ to be enforced;
+- `spec`: describes the desired state when _the thing_ identified by `name` is enforced.
+
+As security profiles are portable,
+the `name` must be globally unique across all security profiles,
+all clusters, and specific to define what to be enforced.
+The schema of `spec` is defined by `name`. When `name` is defined,
+the schema of `spec` is finalized, forever.
+Further change of schema will require a new unique name.
+
+With rules predefined,
+a security profile aware Kubernetes deployment tool should be able to
+download and inspect the selected security profile before creating/upgrading a cluster,
+and setup command line flags, config files properly according to the rules during cluster creation/upgrade.
+
+These rules can also be interpreted as a set of check conditions for tools to validate if the cluster has been setup properly.
+
+Please see [Predefined Rules](#predefined-rules) as a reference of the rules.
+
+As currently, there's no way (or limited) to inspect the host operating system from inside the cluster,
+these rules are only enforced by the deployers (or cluster lifecycle management system).
+
+#### Runtime Rules
+Runtime rules are enforced after the cluster is bootstrapped,
+by _enforcers_ (the controllers deployed with security profiles in the cluster).
+Each rule is defined as an individual Custom Resource, that helps the enforcers define strong typed schemas.
+In a security profile, a rule is a simple reference to a custom resource.
+
+_Alternative design_
+
+Runtime rules can also be defined similarly to bootstrapping rules, in a schemaless way.
+However it loses the support from Kubernetes API to perform schema validation and other functionality,
+even though it's currently very limited for CRDs, in future it may become better.
+
+#### Effectiveness
+Multiple security profiles can exist in the same cluster at the same time,
+but only one can be active (being enforced with `runtime-rules`, not `bootstrapping-rules`).
```

Comment:
I find this confusing - which one would be enforced? (based on the SecurityProfileSelector below?) Is there a hierarchy, if they are in conflict?

Why allow multiple profiles, instead of just forcing users to adopt a different, merged profile?

Reply:
It's possible in the cluster, there are multiple security profiles, like default, saas. User can use SecurityProfileSelector to select one of them. There's no hierarchy, and no conflict. Because only one security profile is selected. It's not allowed to select multiple security profiles.